5calls/app/src/androidTest/java/org/a5calls/android/a5calls/controller/MainActivityErrorTest.java

```diff
@@ -0,0 +1,234 @@
+        }
+
+        @Override
+        public void registerIdleTransitionCallback(ResourceCallback callback) {
+            mResourceCallback = callback;
+        }
+    }
+
+    @Before
+    public void setUp() {
+        // Save original state
+        mOriginalRequestQueue = AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext()).getRequestQueue();
+        mOriginalApi = AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext()).getJsonController();
+
+        // Save original location
+        mOriginalAddress = AccountManager.Instance.getAddress(
+                InstrumentationRegistry.getInstrumentation().getTargetContext());
+
+        // Set a mock location to avoid location prompts
+        AccountManager.Instance.setAddress(
+                InstrumentationRegistry.getInstrumentation().getTargetContext(),
+                "90210");
+
+        // Mark tutorial as seen to bypass onboarding screen
+        AccountManager.Instance.setTutorialSeen(
+                InstrumentationRegistry.getInstrumentation().getTargetContext(),
+                true);
+
+        // Create mock HTTP stack
+        mHttpStack = new MockHttpStack();
+    }
+
+    @After
+    public void tearDown() {
+        // Restore original state
+        AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext())
+                .setRequestQueue(mOriginalRequestQueue);
+        AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext())
+                .setFiveCallsApi(mOriginalApi);
+
+        // Restore original location
+        AccountManager.Instance.setAddress(
+                InstrumentationRegistry.getInstrumentation().getTargetContext(),
+                mOriginalAddress);
+    }
+
+    @Test
+    public void testNetworkErrorDisplaysSnackbar() {
+        // Set up mock to throw network error
+        mHttpStack.setExceptionToThrow(new IOException("Network error"));
+
+        // Create a custom RequestQueue with our mock HTTP stack
+        BasicNetwork basicNetwork = new BasicNetwork(mHttpStack);
+        FakeRequestQueue requestQueue = new FakeRequestQueue(basicNetwork);
+        requestQueue.start();
+
+        // Replace the app's RequestQueue with our mock
+        AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext())
+                .setRequestQueue(requestQueue);
+
+        // Create a new FiveCallsApi with our mock RequestQueue
+        String callerId = AccountManager.Instance.getCallerID(
+                InstrumentationRegistry.getInstrumentation().getTargetContext());
+        FiveCallsApi api = new FiveCallsApi(callerId, requestQueue);
+        AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext())
+                .setFiveCallsApi(api);
+
```

Comment:
a lot of this (lines 125-141) can be refactored into a helper method rather than repeated for each test.